File: src/render.rs
//! AST → Markdown rendering. One deterministic pass: escape source text once,
//! emit structural markers, pad table columns from typed widths. No regexes,
//! no re-parsing of emitted output.

use crate::ast::{Block, Inline, ListItem};

/// Render a whole document: blocks joined by blank lines, blank runs
/// collapsed (double `<br>` can put one inside a paragraph).
pub fn document(blocks: &[Block]) -> String {
    let parts: Vec<String> = blocks
        .iter()
        .map(render_block)
        .filter(|s| !s.is_empty())
        .collect();
    collapse_blank_runs(&parts.join("\n\n"))
}

/// Collapse whitespace between tokens to single spaces (browser whitespace
/// collapsing, applied to rendered strings in whitespace-hostile contexts:
/// table cells, list items, link display).
pub(crate) fn normalize_ws(s: &str) -> String {
    s.split_whitespace().collect::<Vec<_>>().join(" ")
}

/// Render inlines to their Markdown string.
pub(crate) fn inlines_to_string(inlines: &[Inline]) -> String {
    inlines.iter().map(render_inline).collect()
}

/// Trim a multi-line paragraph (one whose `<br>`s became newlines): strip
/// surrounding whitespace on every line so leading spaces from source text
/// nodes don't survive as ragged indentation, while preserving blank lines
/// (from `<br><br>`) between the kept lines. A paragraph that reduces to a
/// single ASCII punctuation char is dropped: it is separator residue, e.g.
/// the ` . ` left between two `<img>` that were removed as decorative media.
pub(crate) fn tidy_paragraph(s: &str) -> String {
    let out = if !s.contains('\n') {
        s.trim().to_string()
    } else {
        s.lines()
            .map(|l| l.trim())
            .collect::<Vec<_>>()
            .join("\n")
            .trim()
            .to_string()
    };
    if out.len() == 1 && out.as_bytes()[0].is_ascii_punctuation() {
        return String::new();
    }
    out
}

fn render_block(block: &Block) -> String {
    match block {
        Block::Paragraph { inlines } => tidy_paragraph(&inlines_to_string(inlines)),
        Block::Heading { level, inlines } => {
            // A heading is a single line; fold any `<br>`-newline to a space.
            let s = normalize_ws(&inlines_to_string(inlines));
            if s.is_empty() {
                String::new()
            } else {
                format!("{} {}", "#".repeat(*level), s)
            }
        }
        Block::Rule => "---".to_string(),
        Block::Code { body } => format!("```\n{body}\n```"),
        Block::Quote { blocks } => {
            let inner: Vec<String> = blocks
                .iter()
                .map(render_block)
                .filter(|s| !s.is_empty())
                .collect();
            if inner.is_empty() {
                return String::new();
            }
            inner
                .join("\n\n")
                .lines()
                .map(|l| {
                    if l.is_empty() {
                        ">".to_string()
                    } else {
                        format!("> {l}")
                    }
                })
                .collect::<Vec<_>>()
                .join("\n")
        }
        Block::List { ordered, items } => render_list(items, *ordered, 0),
        Block::Table { rows } => render_table(rows),
    }
}

fn render_list(items: &[ListItem], ordered: bool, depth: usize) -> String {
    let indent = "  ".repeat(depth);
    let mut lines: Vec<String> = Vec::new();
    for item in items {
        // A list item is one logical line; fold any `<br>`-newline to a space
        // so a continuation doesn't escape the bullet's indentation.
        let text = normalize_ws(&inlines_to_string(&item.inlines));
        if text.is_empty() && item.sub_lists.is_empty() {
            continue;
        }
        let marker = if ordered {
            format!("{}. ", item.number)
        } else {
            "- ".to_string()
        };
        lines.push(format!("{indent}{marker}{text}"));
        for sub in &item.sub_lists {
            // Sub-list already carries its own depth-based indent.
            if let Block::List {
                ordered: sub_ordered,
                items: sub_items,
            } = sub
            {
                for line in render_list(sub_items, *sub_ordered, depth + 1).lines() {
                    lines.push(line.to_string());
                }
            }
        }
    }
    lines.join("\n")
}

fn render_table(rows: &[Vec<Vec<Inline>>]) -> String {
    let ncols = rows.iter().map(|r| r.len()).max().unwrap_or(0);
    if ncols == 0 {
        return String::new();
    }

    // Minimal framing: single space between the pipes and the cells, no
    // column padding. Tables stay narrow and ragged rows don't stretch the
    // frame; terminal concealers don't align raw columns anyway.
    let mut lines: Vec<String> = Vec::new();
    for (i, row) in rows.iter().enumerate() {
        let mut line = String::from("|");
        for c in 0..ncols {
            let cell = match row.get(c) {
                Some(cell) => normalize_ws(&inlines_to_string(cell)),
                None => String::new(),
            };
            line.push(' ');
            line.push_str(&cell);
            line.push_str(" |");
        }
        lines.push(line);
        if i == 0 {
            let mut sep = String::from("|");
            for _ in 0..ncols {
                sep.push_str(" - |");
            }
            lines.push(sep);
        }
    }
    lines.join("\n")
}

fn render_inline(inline: &Inline) -> String {
    match inline {
        Inline::Text(s) => escape_markdown(s),
        Inline::Raw(s) => s.clone(),
        Inline::Code(s) => format!("`{s}`"),
        Inline::Strong(inner) => emphasis(inner, "**"),
        Inline::Emph(inner) => emphasis(inner, "*"),
        Inline::Link { text, href } => format!("[{text}]({href})"),
        Inline::LineBreak => "\n".to_string(),
    }
}

/// Wrap inline content in an emphasis marker (`**` or `*`), keeping any leading
/// or trailing whitespace *outside* the markers. HTML often splits a phrase
/// across adjacent `<strong>`/`<b>` runs where the only separating space lives
/// at a marker boundary (e.g. `<b>Twenty</b><b>&nbsp;minutes</b>`); trimming it
/// away would fuse the words (`**Twenty****minutes**`).
fn emphasis(inner: &[Inline], marker: &str) -> String {
    let s = inlines_to_string(inner);
    // A `<br>` inside the emphasis (`<b>line1<br>line2</b>`) leaves a newline in
    // `s`; wrap each line on its own so the markers never span a line break
    // (`**line1**\n**line2**`), which would otherwise render the literal `**`.
    if s.contains('\n') {
        return s
            .split('\n')
            .map(|line| {
                let t = line.trim();
                if t.is_empty() {
                    String::new()
                } else {
                    format!("{marker}{t}{marker}")
                }
            })
            .collect::<Vec<_>>()
            .join("\n");
    }
    let t = s.trim();
    if t.is_empty() {
        return String::new();
    }
    let lead = if s.starts_with(|c: char| c.is_whitespace()) {
        " "
    } else {
        ""
    };
    let trail = if s.ends_with(|c: char| c.is_whitespace()) {
        " "
    } else {
        ""
    };
    format!("{lead}{marker}{t}{marker}{trail}")
}

/// Escape characters that could create unintended Markdown syntax. Whitespace
/// collapsing and unicode-escape decoding happened at lowering; this only
/// backslash-escapes the structural characters.
fn escape_markdown(s: &str) -> String {
    let mut out = String::with_capacity(s.len() + 4);
    for c in s.chars() {
        match c {
            '\\' | '`' | '*' | '_' | '[' | ']' => {
                out.push('\\');
                out.push(c);
            }
            _ => out.push(c),
        }
    }
    out
}

fn collapse_blank_runs(s: &str) -> String {
    let mut out = s.to_owned();
    while out.contains("\n\n\n") {
        out = out.replace("\n\n\n", "\n\n");
    }
    out
}

File: go/internal/render/render.go
// Package render: AST → Markdown. One deterministic pass: escape source
// text once, emit structural markers. No regexes, no re-parsing of emitted
// output.
package render

import (
	"strings"

	"github.com/stubbe/html-to-md/internal/ast"
)

// Document renders a whole document: blocks joined by blank lines, blank
// runs collapsed.
func Document(blocks []ast.Block) string {
	parts := make([]string, 0, len(blocks))
	for _, b := range blocks {
		if s := block(b); s != "" {
			parts = append(parts, s)
		}
	}
	return collapseBlankRuns(strings.Join(parts, "\n\n"))
}

func collapseBlankRuns(s string) string {
	for strings.Contains(s, "\n\n\n") {
		s = strings.ReplaceAll(s, "\n\n\n", "\n\n")
	}
	return s
}

// NormalizeWs collapses whitespace between tokens to single spaces (browser
// whitespace collapsing, applied in whitespace-hostile contexts).
func NormalizeWs(s string) string {
	words := strings.Fields(s)
	return strings.Join(words, " ")
}

// InlinesToString renders inlines to their Markdown string.
func InlinesToString(inlines []ast.Inline) string {
	var b strings.Builder
	for _, in := range inlines {
		b.WriteString(inline(in))
	}
	return b.String()
}

// TidyParagraph trims an accumulated paragraph: strip surrounding whitespace
// on every line while preserving blank lines from <br><br>. A paragraph that
// reduces to a single ASCII punctuation char is dropped: it is separator
// residue, e.g. the " . " left between two <img> removed as decorative media.
func TidyParagraph(s string) string {
	var out string
	if !strings.Contains(s, "\n") {
		out = strings.TrimSpace(s)
	} else {
		var lines []string
		for l := range strings.SplitSeq(s, "\n") {
			lines = append(lines, strings.TrimSpace(l))
		}
		out = strings.TrimSpace(strings.Join(lines, "\n"))
	}
	if len(out) == 1 && isAsciiPunct(out[0]) {
		return ""
	}
	return out
}

func isAsciiPunct(c byte) bool {
	asciiPunct := "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"
	return strings.IndexByte(asciiPunct, c) >= 0
}

func block(b ast.Block) string {
	switch b.Kind {
	case ast.KindParagraph:
		return TidyParagraph(InlinesToString(b.Inlines))
	case ast.KindHeading:
		// A heading is a single line; fold any <br>-newline to a space.
		s := NormalizeWs(InlinesToString(b.Inlines))
		if s == "" {
			return ""
		}
		return strings.Repeat("#", b.Level) + " " + s
	case ast.KindRule:
		return "---"
	case ast.KindCode:
		return "```\n" + b.Body + "\n```"
	case ast.KindQuote:
		var inner []string
		for _, q := range b.Blocks {
			if s := block(q); s != "" {
				inner = append(inner, s)
			}
		}
		if len(inner) == 0 {
			return ""
		}
		var lines []string
		for l := range strings.SplitSeq(strings.Join(inner, "\n\n"), "\n") {
			if l == "" {
				lines = append(lines, ">")
			} else {
				lines = append(lines, "> "+l)
			}
		}
		return strings.Join(lines, "\n")
	case ast.KindList:
		return list(b.Items, b.Ordered, 0)
	case ast.KindTable:
		return table(b.Rows)
	}
	return ""
}

func list(items []ast.ListItem, ordered bool, depth int) string {
	indent := strings.Repeat("  ", depth)
	var lines []string
	for _, it := range items {
		text := NormalizeWs(InlinesToString(it.Inlines))
		if text == "" && len(it.SubLists) == 0 {
			continue
		}
		marker := "- "
		if ordered {
			marker = itoa(it.Number) + ". "
		}
		lines = append(lines, indent+marker+text)
		for _, sub := range it.SubLists {
			if sub.Kind == ast.KindList {
				for l := range strings.SplitSeq(list(sub.Items, sub.Ordered, depth+1), "\n") {
					lines = append(lines, l)
				}
			}
		}
	}
	return strings.Join(lines, "\n")
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b [12]byte
	i := len(b)
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		return "-" + string(b[i:])
	}
	return string(b[i:])
}

func table(rows ast.Table) string {
	ncols := 0
	for _, r := range rows {
		ncols = max(ncols, len(r))
	}
	if ncols == 0 {
		return ""
	}
	// Minimal framing: single space between the pipes and the cells, no
	// column padding (the terminal concealers don't align raw columns).
	var lines []string
	for i, row := range rows {
		var line strings.Builder
		line.WriteByte('|')
		for c := range ncols {
			cell := ""
			if c < len(row) {
				cell = NormalizeWs(InlinesToString(row[c]))
			}
			line.WriteByte(' ')
			line.WriteString(cell)
			line.WriteString(" |")
		}
		lines = append(lines, line.String())
		if i == 0 {
			var sep strings.Builder
			sep.WriteByte('|')
			for range ncols {
				sep.WriteString(" - |")
			}
			lines = append(lines, sep.String())
		}
	}
	return strings.Join(lines, "\n")
}

func inline(in ast.Inline) string {
	switch in.Kind {
	case ast.IText:
		return escapeMarkdown(in.S)
	case ast.IRaw:
		return in.S
	case ast.ICode:
		return "`" + in.S + "`"
	case ast.IStrong:
		return emphasis(in.Inner, "**")
	case ast.IEmph:
		return emphasis(in.Inner, "*")
	case ast.ILink:
		return "[" + in.LinkText + "](" + in.Href + ")"
	case ast.ILineBreak:
		return "\n"
	}
	return ""
}

// Emphasis wraps inline content in a marker, keeping any leading or trailing
// whitespace *outside* the markers. HTML often splits a phrase across
// adjacent runs where the only separating space lives at a marker boundary.
func emphasis(inner []ast.Inline, marker string) string {
	s := InlinesToString(inner)
	if strings.Contains(s, "\n") {
		// A <br> inside emphasis: wrap each line on its own so markers never
		// span a line break.
		var out []string
		for line := range strings.SplitSeq(s, "\n") {
			t := strings.TrimSpace(line)
			if t == "" {
				out = append(out, "")
			} else {
				out = append(out, marker+t+marker)
			}
		}
		return strings.Join(out, "\n")
	}
	t := strings.TrimSpace(s)
	if t == "" {
		return ""
	}
	lead, trail := "", ""
	if startsWithWs(s) {
		lead = " "
	}
	if endsWithWs(s) {
		trail = " "
	}
	return lead + marker + t + marker + trail
}

func startsWithWs(s string) bool {
	for _, c := range s {
		return isWs(c)
	}
	return false
}

func endsWithWs(s string) bool {
	r := []rune(s)
	if len(r) == 0 {
		return false
	}
	return isWs(r[len(r)-1])
}

func isWs(c rune) bool {
	switch c {
	case ' ', '\t', '\n', '\r', '\v', '\f', 0x85, 0xA0:
		return true
	}
	return false
}

// escapeMarkdown backslash-escapes the structural characters. Whitespace
// collapsing and unicode-escape decoding happened at lowering.
func escapeMarkdown(s string) string {
	var b strings.Builder
	b.Grow(len(s) + 4)
	for _, c := range s {
		switch c {
		case '\\', '`', '*', '_', '[', ']':
			b.WriteByte('\\')
			b.WriteRune(c)
		default:
			b.WriteRune(c)
		}
	}
	return b.String()
}

// Esc escapes the same structural characters (used by the calendar path).
func Esc(s string) string { return escapeMarkdown(s) }
